### PR TITLE
bpf: add new macro __section_entry

### LINKED ISSUE #25616: Tag bpf entrypoints with new `__section_entry` macro
Andrii submitted a patch a while ago to make the Cilium datapath fully conform to libbpf conventions: https://github.com/anakryiko/cilium/commit/e66bc11552ceee4045eac94741f99b9fbcf7989e.

This makes the datapath compile and pass veristat, but it cannot be merged as-is, since it converts tail call maps to BTF map defs that don't have the `id` field needed to hook up tail calls into their respective tail call maps. See https://github.com/cilium/cilium/issues/25005 for more info and ongoing work.

The other part of Andrii's proposed change makes it so entrypoints have section names that conform to libbpf conventions. `from-container`, `to-host` , etc. all need to become either `tc` or `xdp` depending on whether the datapath was built with `skb.h` or `xdp.h`.

I propose to add a new `__section_entry` macro (to go with the existing `__section_tail` and friends) that is declared as  `__section("tc")` by `skb.h` and `__section("xdp")` by `xdp.h`. This also acts as a clear definition of entrypoints vs. tail calls.

### PR DESCRIPTION
**Pull request by Jack-R-lantern**
Please ensure your pull request adheres to the following guidelines:

- [x] For first time contributors, read [Submitting a pull request](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submitting-a-pull-request)
- [x] All code is covered by unit and/or runtime tests where feasible.
- [ ] All commits contain a well written commit description including a title,
      description and a `Fixes: #XXX` line if the commit addresses a particular
      GitHub issue.
- [ ] If your commit description contains a `Fixes: <commit-id>` tag, then
      please add the commit author[s] as reviewer[s] to this issue.
- [x] All commits are signed off. See the section [Developer’s Certificate of Origin](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#dev-coo)
- [ ] Provide a title or release-note blurb suitable for the release notes.
- [ ] Are you a user of Cilium? Please add yourself to the [Users doc](https://github.com/cilium/cilium/blob/main/USERS.md)
- [ ] Thanks for contributing!

Added __section_entry macro for libbpf convetion.
skb.h, xdp.h

Fixes: #25616

### CODE REVIEW

File: bpf/bpf_host.c [hunk 1]
```diff
@@ -1308,7 +1308,7 @@ handle_srv6(struct __ctx_buff *ctx)
  * - L2 announcements are enabled, or
  * - WireGuard's host-to-host encryption and BPF NodePort are enabled
  */
-__section("from-netdev")
+__section_entry("from-netdev")
```
Comment (ti-mo) [inline bpf/bpf_host.c]: There's no point in retaining the section names, let's make these `__section_entry int cil_...`

File: bpf/include/bpf/ctx/skb.h [hunk 2]
```diff
@@ -4,6 +4,10 @@
 #ifndef __BPF_CTX_SKB_H_
 #define __BPF_CTX_SKB_H_
 
+#ifndef __section_entry
```
Comment (ti-mo) [inline bpf/include/bpf/ctx/skb.h]: Do we need to gate the define? I'd rather have a compiler error if this gets defined more than once.

**Review by ti-mo — CHANGES_REQUESTED** [review]

File: bpf/include/bpf/ctx/skb.h (hunk 2 above)
Comment (Jack-R-lantern) [reply bpf/include/bpf/ctx/skb.h]: @ti-mo Is it okay if the error message is '__section_entry is already defined'?
[skb.h](https://github.com/cilium/cilium/pull/26123/files#diff-8c2d304a3199e3dcfc93bd42ac49ec4aa74b466da8d1d41ea1c55ec8b01922d1R7-R9)
[xdp.h](https://github.com/cilium/cilium/pull/26123/files#diff-a8e03cca23ef430330659f761b5118e3f0d64fbd350d8e768b1e936aef250145R10-R12)

File: bpf/bpf_host.c (hunk 1 above)
Comment (Jack-R-lantern) [reply bpf/bpf_host.c]: Fixed [skb __section_entry](https://github.com/cilium/cilium/pull/26123/files#diff-8c2d304a3199e3dcfc93bd42ac49ec4aa74b466da8d1d41ea1c55ec8b01922d1R11)
Fixed [xdp __section_entry](https://github.com/cilium/cilium/pull/26123/files#diff-a8e03cca23ef430330659f761b5118e3f0d64fbd350d8e768b1e936aef250145R14)
all section name remove

File: bpf/include/bpf/ctx/skb.h [hunk 3]
```diff
@@ -4,6 +4,12 @@
 #ifndef __BPF_CTX_SKB_H_
 #define __BPF_CTX_SKB_H_
 
+#ifdef __section_entry
```
Comment (ti-mo) [inline bpf/include/bpf/ctx/skb.h]: No need for this, the compiler already throws this including the location where it was defined previously

**Review by ti-mo — CHANGES_REQUESTED** [review]

File: bpf/include/bpf/ctx/skb.h (hunk 3 above)
Comment (Jack-R-lantern) [reply bpf/include/bpf/ctx/skb.h]: Thank you for the review, I removed the error directive xdp.h, skb.h

**Review by ti-mo — APPROVED** [review]
Thanks!

**Comment by ti-mo** [comment]
I'd like to ship this in 1.14, so I've kicked off tests and tagged this as a release blocker as a reminder to revisit this at the end of the week.

**Comment by ldelossa** [comment]
@ti-mo is it worth backing-porting this to our stable branches?

**Review by ldelossa — APPROVED** [review]
Changes look good to me.

**Comment by ti-mo** [comment]
@Jack-R-lantern I've squashed your commits into a single one and added a small fixup that was causing the ginkgo tests to fail.

**Pull request merged**